### ISSUE TITLE
Using ResourceTag with iam:PassRole conditions prevents env interaction

### ISSUE THREAD

**Issue by miketheman** [aws/copilot-cli#2160]
(originally reported as part of a [Discussion thread](https://github.com/aws/copilot-cli/discussions/2140))

After deploying to an environment with `copilot svc deploy --name app` I was unable to remove the `env` that was created, getting this IAM response:

> Are you sure you want to delete environment load from application app1? Yes
> ✘ Failed to delete environment load from application app1.
> ✘ delete environment load stack: delete stack app1-load: AccessDenied: User: arn:aws:sts::[REDACTED]:assumed-role/app1-load-EnvManagerRole/1617489726744667000 is not authorized to perform: iam:PassRole on resource: arn:aws:iam::[REDACTED]:role/app1-load-CFNExecutionRole
> status code: 403, request id: 0a03c9fd-6fb8-436b-92ec-60144cf2a58e

After some further debugging, I found:

The iam:PassRole permission in the `EnvManagerRole` policy specifies:

```json

{
    "Condition": {
        "StringEquals": {
            "iam:ResourceTag/copilot-application": "[MY APP NAME]",
            "iam:ResourceTag/copilot-environment": "[MY ENV NAME]"
        }
    },
    "Action": [
        "iam:GetRole",
        "iam:PassRole"
    ],
    "Resource": "*",
    "Effect": "Allow",
    "Sid": "GetAndPassCopilotRoles"
},
```

When looking at the `CFNExecutionRole`, there are no Tags associated with the role - so the condition could not be satisfied.

I then tried to manually add both conditions to the Tags on `CFNExecutionRole`, but got the same error response.

Reading further, the [AWS IAM Docs](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_tags.html#access_tags_control-resources
) also specifically call out to NOT use ResourceTags for PassRole:

> Do not use the ResourceTag condition key in a policy with the iam:PassRole action.
> You cannot use the tag on an IAM role to control access to who can pass that role.

I ended up manually removing the condition from the policy in the IAM console to be able to complete the env removal.

So it seems like there's a problem with how we're intending on restricting access to the PassRole condition?

**Comment by iamhopaul123** [comment]
Hello @miketheman. Thank you for documenting this unexpected behavior!

From the error i see
```
User: arn:aws:sts::[REDACTED]:assumed-role/app1-load-EnvManagerRole/1617489726744667000 is not authorized to perform: iam:PassRole on resource: arn:aws:iam::[REDACTED]:role/app1-load-CFNExecutionRole
```
It seems like the `envManagerRole` cannot perform `iam:PassRole` to `CFNExecutionRole`. So I guess for whatever reason the `app1-load-CFNExecutionRole` isn't tagged correctly. I think that's the reason why it failed. Basically according to my test you have to add the tag in this order:
```
copilot-environment prod
copilot-application couch-demo
```
If you swap those two tags upside down then it's not gonna work. I think it is a very strange behavior on IAM side and that's probably part of the reason why they don't recommend to use `ResourceTag` on `iam:PassRole`. Please keep tracking this issue. We should probably follow the IAM recommendation to ditch the condition. Again, thank you so much for reporting this issue.

**Comment by iamhopaul123** [comment]
Hello @miketheman. Also do you know the reason why those tags for the IAM roles did not exist? Because even if we update the permission, without those tags Copilot still won't work for the other commands, since there's no limitation for `ResourceTag` with the other actions.

**Comment by miketheman** [comment]
Hey @iamhopaul123 !

>  do you know the reason why those tags for the IAM roles did not exist?

I tried again today, and with a basic `copilot init`, the autogenerated `test` environment doesn't have the Resource Tags either.

The Cloudformation **Stack** shows the correct Tags that are supposed to be inherited by any resource created by the CFN template. Other resources, like VPC, show these two tags, as well as `aws:cloudformation:stack-name`, `aws:cloudformation:stack-id`, and `aws:cloudformation:logical-id` - very in line with CFN expected behavior.

Looks like this may be related? https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/328

Does this not reproduce for you?

**Comment by miketheman** [comment]
And to the Tag order part - I logged into the account's IAM privileged user account (yay safety!) and manually applied the two tags to the `CFNExecutionRole` in the order you specified.
I was then able to successfully delete the stack with `copilot app delete` (I'm guessing that `copilot env delete ...` would have worked as well) - so at least that is a possible.

**Comment by miketheman** [comment]
I continued to test this, and it (should have been obvious, but wasn't 🤦 ) prevents any interactions that would use PassRole, like `copilot svc deploy....` too. I've updated the description to reflect that finding.
Similarly, adding the Resources Tags manually (in the correct order) unblocks the execution.

**Comment by iamhopaul123** [comment]
> Does this not reproduce for you?

Unfortunately I am not able to reproduce this one and it seems like you are the first user reporting this issue (IAM roles produced by CFN do not inherit tags from CFN). Does it happen every time when you create an environment? Also thank you so much for your efforts for investigating this!

**Comment by miketheman** [comment]
@iamhopaul123 Did you happen to dig into the linked issue from the cloudformation link above?

My understanding is that my user creates a role, and that role creates new IAM resources - and based on the CFN bug link, a role creating IAM resources via CFN won't tag-on-create.

It's pretty consistent - I create the app, then I create an env, and then we see the issue. If there's some other debugging I can provide, let me know, and I'll do what I can to produce!

**Comment by iamhopaul123** [comment]
Oh I see what you meant. Did you create the environment using temporary credentials of an IAM role?

**Comment by miketheman** [comment]
No, the credentials are (I think) long-lived IAM user access key/secret - I don't think the linked cloudformation issue specifies that it applies to temp credentials, does it?

**Comment by Lou1415926** [comment]
Hello @miketheman ! Sorry for the delayed follow-up. I am revisiting this issue, and am wondering if it's possible that you have scoped down your root account permission such that it lacks the `iam:TagRole`? For example, 
```
copilot env init --name test --profile env-profile
```
Is it possible that the policy for `env-profile` doesn't have the `iam:TagRole`, which led to the CFNExecution role not properly tagged?

**Comment by miketheman** [comment]
Hey @Lou1415926 Thanks for following up, and apologies for not responding sooner!

I actually ran into this today again, which reminded me of the open issue.

For clarity - this isn't using the AWS account root user, rather a slimmed down "developer" account, with far more restrictive permissions.

> Is it possible that the policy for `env-profile` doesn't have the `iam:TagRole`, which led to the CFNExecution role not properly tagged?

Indeed, the user did not have `iam:TagRole` allowed, which I've corrected and everything appears to work 🎉 
To help myself and others in the future:

```json
...
            "Effect": "Allow",
            "Action": "iam:TagRole",
            "Resource": "arn:aws:iam::123456789012:role/*"
...
```

I think this points out the (continued) need for clarifying what the initial minimal permissions are necessary to accomplish a Copilot setup and deployment (ref #2615).

It's also something that should be tested as a prerequisite so that the user doesn't end up in a state where they have created something they cannot manage/remove without surfacing this issue.

Thanks for the helpful pointer, and let me know if I can help improve this experience!
